tool/tsh/common/vnet_darwin.go

```diff
@@ -44,7 +47,25 @@ func (c *vnetCommand) run(cf *CLIConf) error {
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	return trace.Wrap(vnet.Run(cf.Context, appProvider))
+
+	processManager, err := vnet.SetupAndRun(cf.Context, appProvider)
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		<-cf.Context.Done()
+		processManager.Close()
+	}()
+
+	err = processManager.Wait()
+	if errors.Is(err, context.Canceled) {
+		return nil
+	}
+	return trace.Wrap(err)
```

Comment:
yeah again I think it's fine to return the error even if we got killed by ctrl-c. If you kill any other long-running command like `tsh login` (waiting for password/mfa) it returns an error and that's fine

this also avoids just returning nil if we some returned context.Canceled without the user actually trying to kill vnet with ctrl-c

```suggestion
	return trace.Wrap(processManager.Wait())
```

Reply:
I think you're generally right and I was going to bring up `tsh login` as an example too – it just got stuck at reading my MFA key and Ctrl+C returned context canceled + a useful user message as to where the code got stuck.

I'm trying to build a good mental model for this. For example, `tsh proxy app` emits no message on context canceled after the proxy is up and returns 0 exit code. `tail` emits no message on Ctrl+C, but the exit code is 130 (command exited through sigint).

Originally, I added checks for context.Canceled because it was annoying to see "context canceled" when I wanted to close VNet, it seemed like a bad UX.

Now I realize that maybe how we depend on ctx and how we always print the error message is wrong? That is, VNet should somehow wait for SIGINT and if it's received while VNet is idle, "context canceled" should not be presented to the user. It should be emitted though while VNet is waiting for something to happen.

For now I'll just always return the error.